api/v1/controllers/collectors.js

```diff
@@ -159,44 +214,91 @@ function deregisterCollector(req, res, next) {
-   */
-
-  /*
-   * TODO populate generatorsAdded
-   * - look up any new generators assigned to this collector since the last
-   *   heartbeat
-   * - reEncrypt context values using reEncryptSGContextValues function
-   */
-
-  /*
-   * TODO populate generatorsDeleted
-   * - look up any generators UNassigned from this collector since the last
-   *   heartbeat
-   */
-
-  /*
-   * TODO populate generatorsUpdated
-   * - for generators which were already assigned to this collector, look up
-   *   any changes made to the generator since the last heartbeat
-   * - reEncrypt context values using reEncryptSGContextValues function
-   */
-
-  res.status(httpStatus.OK).json(retval);
+  //verify token
+  jwtUtil.verifyCollectorToken(req)
+  .then(() => jwtUtil.getTokenDetailsFromRequest(req))
+  .then((details) => tokenName = details.username)
+
+  //find collector object
+  .then(() => u.findByKey(helper, req.swagger.params))
+  .then((o) => {
+    keyName = o.name;
+
+    //validate collector
+    if (keyName !== tokenName) {
+      throw new apiErrors.ForbiddenError({
+        explanation: 'Token does not match the specified collector',
+      });
+    } else if (o.status !== 'Running' && o.status !== 'Paused') {
+      throw new apiErrors.ForbiddenError({
+        explanation: `Collector must be running or paused. Status: ${o.status}`,
+      });
+    }
+
+    //setup retval
+    if (heartbeatUtils.collectorMap[o.name]) {
+      retval.generatorsAdded = heartbeatUtils.collectorMap[o.name].added;
+      retval.generatorsDeleted = heartbeatUtils.collectorMap[o.name].deleted;
+      retval.generatorsUpdated = heartbeatUtils.collectorMap[o.name].updated;
+      delete heartbeatUtils.collectorMap[o.name];
+    }
+
+    //set lastHeartbeat
+    o.set('lastHeartbeat', timestamp);
+
+    //update metadata
+    const changedConfig = req.body.collectorConfig;
+    if (changedConfig) {
+      if (changedConfig.osInfo) {
+        const osInfo = o.osInfo ? o.osInfo : {};
+        Object.assign(osInfo, changedConfig.osInfo);
+        o.set('osInfo', osInfo);
+      }
+
+      if (changedConfig.processInfo) {
+        const processInfo = o.processInfo ? o.processInfo : {};
+        Object.assign(processInfo, changedConfig.processInfo);
+        o.set('processInfo', processInfo);
+      }
+
+      if (changedConfig.version) {
+        o.set('version', changedConfig.version);
+      }
+    }
+
+    return o.save();
+  })
+
+  //re-encrypt context values for added and updated generators
+  .then(() => Promise.all(
+    retval.generatorsAdded.map((sg) =>
+      attachTemplate(sg)
+      .then((sg) => reEncryptSGContextValues(sg, authToken, timestamp))
+    )
+  ))
+  .then((added) => retval.generatorsAdded = added)
+  .then(() => Promise.all(
+    retval.generatorsUpdated.map((sg) =>
+      attachTemplate(sg)
+      .then((sg) => reEncryptSGContextValues(sg, authToken, timestamp))
+    )
+  ))
+  .then((updated) => retval.generatorsUpdated = updated)
```

Comment:
Are you trying to return something here ? This reads something like `return retval.generatorsUpdated = updated`. Same comment in line 290.

Reply:
It doesn't matter, does it? Nothing is being used in the next "then" so it's just a formatting choice.

Reply:
By not enclosing `retval.generatorsUpdated = updated` within the curly brackets you are trying to return the assignment operator.

Reply:
Yeah, I know. Does it matter?
I'm happy to add brackets if you think it makes it more readable.

Reply:
No problem. You can leave it this way.